src/python/bot/fuzzers/syzkaller/fuzzer.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""syzkaller fuzzer."""
+from __future__ import absolute_import
+from base import utils
+from bot.fuzzers import builtin
+from bot.fuzzers import utils as fuzzer_utils
+from bot.fuzzers.syzkaller import constants
+from bot.fuzzers.syzkaller import generate_config
+from builtins import object
+from system import environment
+from system import new_process
+import copy
+import re
+import tempfile
+
+# Regex to find testcase path from a crash.
+KASAN_CRASH_TESTCASE_REGEX = (r'.*Test unit written to\s*'
+                              r'(Read|Write) of .*')
+
+
+def get_arguments(fuzzer_path):
+  """Get arguments for a given fuzz target."""
+  del fuzzer_path
+  build_dir = environment.get_value('BUILD_DIR')
+  device_serial = environment.get_value('ANDROID_SERIAL')
+  json_config_path = '/tmp/' + device_serial + '/config.json'
+  generate_config.run(
+      serial=device_serial,
+      work_dir_path='/tmp/syzkaller',
+      binary_path=build_dir + '/syzkaller',
+      vmlinux_path='/tmp/syzkaller/vmlinux',
+      config_path=json_config_path,
+      kcov=True,
+      reproduce=False)
+  arguments = ['--config', json_config_path]
+  return arguments
+
+
+def get_runner(fuzzer_path):
+  """Return a suzkaller runner object."""
+  build_dir = environment.get_value('BUILD_DIR')
+  return SyzkallerRunner(fuzzer_path, build_dir)
+
+
+class Syzkaller(builtin.EngineFuzzer):
+  """Builtin syzkaller fuzzing engine."""
+
+  def generate_arguments(self, fuzzer_path):
+    """Generate arguments for fuzzer using .options file or default values."""
+    return ' '.join(get_arguments(fuzzer_path))
+
+
+class SyzkallerRunner(new_process.ProcessRunner):
+  """Syzkaller runner."""
+
+  def __init__(self, executable_path, default_args=None):
+    """Inits the SyzkallerRunner.
+
+    Args:
+      executable_path: Path to the fuzzer executable.
+      default_args: Default arguments to always pass to the fuzzer.
+    """
+    super(SyzkallerRunner, self).__init__(
+        executable_path=executable_path, default_args=None)
```

Comment:
Could you please clarify what is actually being run here? Is this some wrapper binary which actually invokes syzkaller?

Reply:
syzkaller consists of multiple binaries but the one that orchestrates all the work is syz-manager. Syz-manager runs and pushes a binary on the android device that will actually do the device fuzzing.

Reply:
In that case, how are "fuzz targets" discovered by CF? We detect libFuzzer targets like so: 

https://github.com/google/clusterfuzz/blob/a53d92adb59155a5a5fa194b745389aa421394d8/src/python/bot/fuzzers/utils.py#L33

Reply:
As mentioned in my other note, syzkaller is the engine and the fuzzer (unlike libfuzzer).

Reply:
Yes, so this won't work as is, as CF expects to find a target binary and pass the path through to here. Add a TODO to make that work.

Reply:
Added. But note that using local testing the binary was downloaded and unpacked and run successfully without any change here.